File: prism/runtime/dlt_runner/events.py
"""Structured stdout event emitter for the prism dlt runner.

Events are JSON objects, one per line, written to a configurable stream
(stdout in production, StringIO in tests). See ADR-006 and the design spec
section on IPC for the full event vocabulary.
"""

from __future__ import annotations

import json
import sys
from typing import IO, Any


class Emitter:
    def __init__(self, stream: IO[str] | None = None) -> None:
        self._stream = stream if stream is not None else sys.stdout

    def _emit(self, payload: dict[str, Any]) -> None:
        self._stream.write(json.dumps(payload, separators=(",", ":")) + "\n")
        self._stream.flush()

    def source_start(self, source: str) -> None:
        self._emit({"event": "source.start", "source": source})

    def source_end(self, source: str, entities: int, duration_ms: int) -> None:
        self._emit({
            "event": "source.end", "source": source,
            "entities": entities, "duration_ms": duration_ms,
        })

    def entity_start(self, entity: str) -> None:
        self._emit({"event": "entity.start", "entity": entity})

    def entity_progress(self, entity: str, rows: int) -> None:
        self._emit({"event": "entity.progress", "entity": entity, "rows": rows})

    def entity_end(self, entity: str, rows: int, load_id: str, files: int) -> None:
        self._emit({
            "event": "entity.end", "entity": entity,
            "rows": rows, "load_id": load_id, "files": files,
        })

    def error(self, entity: str, kind: str, message: str) -> None:
        self._emit({
            "event": "error", "entity": entity,
            "kind": kind, "message": message,
        })

File: prism/internal/events/events.go
// prism/internal/events/events.go
// Package events models the JSONL event stream emitted by the prism dlt runner.
// See ADR-006 and the design spec section on IPC.
package events

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"strings"
)

// Event represents a single structured event emitted by the Python runner over stdout.
type Event struct {
	Event    string `json:"event"`
	Source   string `json:"source,omitempty"`
	Entity   string `json:"entity,omitempty"`
	Rows     int    `json:"rows,omitempty"`
	LoadID   string `json:"load_id,omitempty"`
	Files    int    `json:"files,omitempty"`
	Entities int    `json:"entities,omitempty"`
	Duration int    `json:"duration_ms,omitempty"`
	Kind     string `json:"kind,omitempty"`
	Message  string `json:"message,omitempty"`
}

// Parse reads JSONL events from r, calling fn for each. Non-JSON lines are
// reported via fn as Event{Event:"runner.warn", Message: line}. Returns the
// first hard error (read failure or fn-returned error).
func Parse(r io.Reader, fn func(Event) error) error {
	sc := bufio.NewScanner(r)
	sc.Buffer(make([]byte, 0, 64*1024), 1<<20)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			continue
		}
		var e Event
		if err := json.Unmarshal([]byte(line), &e); err != nil {
			if err := fn(Event{Event: "runner.warn", Message: line}); err != nil {
				return err
			}
			continue
		}
		if err := fn(e); err != nil {
			return err
		}
	}
	if err := sc.Err(); err != nil {
		return fmt.Errorf("read events: %w", err)
	}
	return nil
}
